Add Bitdefender (BEST) for Linux support

OscarBeiro opened the pull request:
As mentioned in #458, our best effort on supporting Bitdefender Endpoint Security Tools (BEST) for Linux

OscarBeiro committed:
WIP Add Bitdefender (BEST) for Linux support

OscarBeiro commented:
I'm getting these messages, and the antivirus partial inventory is only showing the Asset information :(

```
Subroutine isEnabled redefined at /usr/share/glpi-agent/lib/GLPI/Agent/Task/Inventory/Linux/AntiVirus/Bitdefender.pm line 13.
Subroutine doInventory redefined at /usr/share/glpi-agent/lib/GLPI/Agent/Task/Inventory/Linux/AntiVirus/Bitdefender.pm line 17.
Subroutine _getBitdefenderInfo redefined at /usr/share/glpi-agent/lib/GLPI/Agent/Task/Inventory/Linux/AntiVirus/Bitdefender.pm line 35.
Subroutine isEnabled redefined at /usr/share/glpi-agent/lib/GLPI/Agent/Task/Inventory/Linux/AntiVirus/Bitdefender.pm line 13.
Subroutine doInventory redefined at /usr/share/glpi-agent/lib/GLPI/Agent/Task/Inventory/Linux/AntiVirus/Bitdefender.pm line 17.
Subroutine _getBitdefenderInfo redefined at /usr/share/glpi-agent/lib/GLPI/Agent/Task/Inventory/Linux/AntiVirus/Bitdefender.pm line 35.
[debug] module GLPI::Agent::Task::Inventory::Linux::AntiVirus::Bitdefender disabled: failure to load (GLPI/Agent/Task/Inventory/Linux/AntiVirus/Bitdefender.pm did not return a true value at /usr/share/glpi-agent/lib/GLPI/Agent/Task/Inventory.pm line 390.
```
Any help is appreciated.

g-bougard commented on lib/GLPI/Agent/Task/Inventory/Linux/AntiVirus/Bitdefender.pm:
```diff
@@ -0,0 +1,81 @@
+use GLPI::Agent::Tools;
+
+sub isEnabled {
+    return canRun('bduitool');
```
Are you sure `bduitool` is in the PATH ? If not, you need to set a full path to the expected one.

OscarBeiro replied:
You're right. It is not in the path by default. Right path is 
/opt/bitdefender-security-tools/bin/bduitool

I'll try to edit it

OscarBeiro replied:
This one is also done.

g-bougard commented on lib/GLPI/Agent/Task/Inventory/Linux/AntiVirus/Bitdefender.pm:
```diff
@@ -0,0 +1,81 @@
+
+sub _getBitdefenderInfo {
+    my (%params) =  (
+        command => '/opt/bitdefender-security-tools/bin/bduitool get ps',
```
If the command is in the PATH, you don't need to use a full path. Indeed, you must be coherent with the `canRun()` call made in `isEnabled()` method.

OscarBeiro replied:
A full path is needed.

g-bougard commented on lib/GLPI/Agent/Task/Inventory/Linux/AntiVirus/Bitdefender.pm:
```diff
@@ -0,0 +1,81 @@
+        @_
+    );
+
+    my $bduitool_output = getAllLines(%params)
+        or return;
```
Better use an array here to simplify a little below.
```suggestion
    my @output = getAllLines(%params)
        or return;
```

g-bougard commented on lib/GLPI/Agent/Task/Inventory/Linux/AntiVirus/Bitdefender.pm:
```diff
@@ -0,0 +1,81 @@
+    my $new_update_available = '';
+    my $new_security_content_available = '';
+
+    foreach my $line (split("\n", $bduitool_output)) {
```
The simplification is here:
```suggestion
    foreach my $line (@output) {
```

g-bougard commented on lib/GLPI/Agent/Task/Inventory/Linux/AntiVirus/Bitdefender.pm:
```diff
@@ -0,0 +1,81 @@
+    my $new_security_content_available = '';
+
+    foreach my $line (split("\n", $bduitool_output)) {
+        if ($line =~ /^Product version:/) {
+            ($product_version = $line) =~ s/^Product version:\s+//;
+        } elsif ($line =~ /^Engines version:/) {
+            ($engines_version = $line) =~ s/^Engines version:\s+//;
+        } elsif ($line =~ /^Antimalware status:/) {
+            ($antimalware_status = $line) =~ s/^Antimalware status:\s+//;
+        } elsif ($line =~ /^New product update available:/) {
+            ($new_update_available = $line) =~ s/^New product update available:\s+//;
+        } elsif ($line =~ /^New security content available:/) {
+            ($new_security_content_available = $line) =~ s/^New security content available:\s+//;
+        }
```
The logic here is too complex, use better something like:
```suggestion
        my ($key, $value) = $line =~ /^([^:]+):\s+(.*)$/
            or next;
        if ($key eq "Product version") {
            $product_version = $value;
        } elsif ($key eq "Engines version") {
            $engines_version = $value;
        } elsif ($key eq "Antimalware status") {
            $antimalware_status = $value;
        } elsif ($key eq "New product update available") {
            $new_update_available = $value;
        } elsif ($key eq "New security content available") {
            $new_security_content_available = $value;
        }
```

Eventually, you can even avoid to use temporary variable and directly update the hash to return. The following could then replace l.44 to 80:
```perl
    my $av = {
        "name" => 'Bitdefender Endpoint Security Tools (BEST) for Linux',
        "manufacturer" => 'Bitdefender',
        "company" => 'Bitdefender',
        "uptodate" => 1,
    };

    foreach my $line (@output) {
        my ($key, $value) = $line =~ /^([^:]+):\s+(.+)$/
            or next;
        if ($key eq "Product version") {
            $av->{version} = $value;
        } elsif ($key eq "Engines version") {
            $av->{base_version} = $value;
        } elsif ($key eq "Antimalware status") {
            $av->{enabled} = $value eq "On" ? 1 : 0;
        } elsif ($key =~ /New (product update|security content) available/) {
            $av->{uptodate} = 0 if $value ne "no";
        }
    }

    return $av;
```

OscarBeiro replied:
This is too complex for my limited knowledge. I'll accept the suggestion :)

g-bougard commented on lib/GLPI/Agent/Task/Inventory/Linux/AntiVirus/Bitdefender.pm:
```diff
@@ -0,0 +1,81 @@
+        "uptodate" => $uptodate,
+        "expiration" => ''
+    };
+}
```
You always need to finish your module by setting return value to 1. This may explain your "redefined" errors.
```suggestion
}

1;

```

g-bougard commented on lib/GLPI/Agent/Task/Inventory/Linux/AntiVirus/Bitdefender.pm:
```diff
@@ -0,0 +1,81 @@
+use parent 'GLPI::Agent::Task::Inventory::Module';
+
+use UNIVERSAL::require;
+use Cpanel::JSON::XS;
```
You should remove this requirement if you don't need json support.

```suggestion
```

g-bougard reviewed (changes requested):
Hi @OscarBeiro 
thank you for your contribution.
Can you try to follow my suggestions ?

g-bougard commented:
Also looking at your output sample:
```
Product version: 7.0.3.2239
Last product update: 2023-08-24 at 10:44:23 GMT+2
New product update available: no
Engines version: 7.95171
Last security content update: 2023-08-24 at 11:44:37 GMT+2
New security content available: no
Installed scan type: Full
Installed scan type fallback: None
Currently used scan type: Full
Features:
  - Antimalware status: On
  - NetworkMonitor status: On
```

It seems "base_creation" can be set parsing "Last security content update" value. Maybe with this code sample:
```perl
        } elsif ($key eq "Last security content update" && $value =~ /^(\d{4}-\d+-\d+) at/) {
            $av->{base_creation} = $1;
```

OscarBeiro committed:
Update lib/GLPI/Agent/Task/Inventory/Linux/AntiVirus/Bitdefender.pm

Remove JSON requirement

Co-authored-by: Guillaume Bougard <gbougard@teclib.com>

OscarBeiro committed:
Update lib/GLPI/Agent/Task/Inventory/Linux/AntiVirus/Bitdefender.pm

Co-authored-by: Guillaume Bougard <gbougard@teclib.com>

OscarBeiro committed:
Update lib/GLPI/Agent/Task/Inventory/Linux/AntiVirus/Bitdefender.pm

Co-authored-by: Guillaume Bougard <gbougard@teclib.com>

OscarBeiro committed:
Suggested simplified logic.

Update with additional suggestions if possible.

Co-authored-by: Guillaume Bougard <gbougard@teclib.com>

OscarBeiro committed:
Fix redefined errors

Missing 1 value.

Co-authored-by: Guillaume Bougard <gbougard@teclib.com>

OscarBeiro committed:
Add missing path as suggested

Signed-off-by: Oscar B. <oscar@ticgal.com>

OscarBeiro committed:
add parameter to bduitool

Running bduitool without parameters returns an error it was also logging a: 
Use of uninitialized value in concatenation (.) or string at /usr/share/glpi-agent/lib/GLPI/Agent/Task/Inventory/Linux/AntiVirus/Bitdefender.pm line 29.


Signed-off-by: Oscar B. <oscar@ticgal.com>

OscarBeiro commented:
> Also looking at your output sample:
> 
> ```
> Product version: 7.0.3.2239
> Last product update: 2023-08-24 at 10:44:23 GMT+2
> New product update available: no
> Engines version: 7.95171
> Last security content update: 2023-08-24 at 11:44:37 GMT+2
> New security content available: no
> Installed scan type: Full
> Installed scan type fallback: None
> Currently used scan type: Full
> Features:
>   - Antimalware status: On
>   - NetworkMonitor status: On
> ```
> 
> It seems "base_creation" can be set parsing "Last security content update" value. Maybe with this code sample:
> 
> ```perl
>         } elsif ($key eq "Last security content update" && $value =~ /^(\d{4}-\d+-\d+) at/) {
>             $av->{base_creation} = $1;
> ```

Good one!

OscarBeiro committed:
Add Base Creation 

Signed-off-by: Oscar B. <oscar@ticgal.com>

OscarBeiro commented:
Still not getting any useful return :(

I have this error, but it seems related to debug2 not having a couple of variables initializated.

Use of uninitialized value in concatenation (.) or string at /usr/share/glpi-agent/lib/GLPI/Agent/Task/Inventory/Linux/AntiVirus/Bitdefender.pm line 29.

So currently:

```
   "content": {
      "antivirus": [
         {}
      ],
```

g-bougard commented on lib/GLPI/Agent/Task/Inventory/Linux/AntiVirus/Bitdefender.pm:
```diff
@@ -0,0 +1,87 @@
+use GLPI::Agent::Tools;
+
+sub isEnabled {
+    return canRun('/opt/bitdefender-security-tools/bin/bduitool about');
```
Don't put an option here. The agent won't run the command here but will check if the given command is runable (see [canRun() API](https://github.com/glpi-project/glpi-agent/blob/3acd88fb7d7df84d5231f09b9948c75d882fd059/lib/GLPI/Agent/Tools.pm#L559))... and in your case, this would be the full path to a command which contains a space and then doesn't exist. And so the module will never be enabled.

OscarBeiro replied:
OK. Done. 
Merci

g-bougard commented on lib/GLPI/Agent/Task/Inventory/Linux/AntiVirus/Bitdefender.pm:
```diff
@@ -0,0 +1,87 @@
+    my $engines_version = '';
+    my $enabled = '';
+    my $antimalware_status = '';
+    my $new_update_available = '';
+    my $new_security_content_available = '';
+
+    foreach my $line (@output) {
+        my ($key, $value) = $line =~ /^([^:]+):\s+(.*)$/
+            or next;
+        if ($key eq "Product version") {
+            $product_version = $value;
+        } elsif ($key eq "Engines version") {
+            $engines_version = $value;
+        } elsif ($key eq "Antimalware status") {
+            $antimalware_status = $value;
+        } elsif ($key eq "New product update available") {
+            $new_update_available = $value;
+        } elsif ($key eq "New security content available") {
+            $new_security_content_available = $value;
+        } elsif ($key eq "Last security content update" && $value =~ /^(\d{4}-\d+-\d+) at/) {
+            $av->{base_creation} = $1;
+        }
+    }
+
+    $enabled = ($antimalware_status eq 'On') ? 1 : 0;
+
+    # Set "uptodate" to 1 if both "new product update" and "new security content" are 'no'
+    my $uptodate = ($new_update_available eq 'no' && $new_security_content_available eq 'no') ? 1 : 0;
+
+    return {
+        "name" => 'Bitdefender Endpoint Security Tools (BEST) for Linux',
+        "manufacturer" => 'Bitdefender',
+        "company" => 'Bitdefender',
+        "version" => $product_version,
+        "base_version" => $engines_version,
+        "base_creation" => '',
+        "enabled" => $enabled,
+        "uptodate" => $uptodate,
+        "expiration" => ''
+    };
```
Sorry for my too complex suggestions.

As `$av` is not existing l.64, you need to define it before. Ideally, here is the code I indeed expected with `$av` defined as a `HASH` with few preset values.

```suggestion
    my $av = {
        "name" => 'Bitdefender Endpoint Security Tools (BEST) for Linux',
        "manufacturer" => 'Bitdefender',
        "company" => 'Bitdefender',
        "uptodate" => 1,
    };

    foreach my $line (@output) {
        my ($key, $value) = $line =~ /^([^:]+):\s+(.+)$/
            or next;
        if ($key eq "Product version") {
            $av->{version} = $value;
        } elsif ($key eq "Engines version") {
            $av->{base_version} = $value;
        } elsif ($key eq "Antimalware status") {
            $av->{enabled} = $value eq "On" ? 1 : 0;
        } elsif ($key =~ /New (product update|security content) available/) {
            # Set "uptodate" to 0 if one of "new product update available" or "new security content available" is not "no"
            $av->{uptodate} = 0 if $value ne "no";
        } elsif ($key eq "Last security content update" && $value =~ /^(\d{4}-\d+-\d+) at/) {
            $av->{base_creation} = $1;
        }
    }

    return $av;
```

g-bougard reviewed (changes requested):
We are getting it better, but as you saw, there's still few mistake.

OscarBeiro committed:
Simplified parsing

Co-authored-by: Guillaume Bougard <gbougard@teclib.com>

OscarBeiro committed:
Remove parameter

Signed-off-by: Oscar B. <oscar@ticgal.com>

OscarBeiro commented:
Sorry 
Still no result. Relevant log:
```
[debug] Running GLPI::Agent::Task::Inventory::Linux::AntiVirus::Bitdefender
[debug2] executing /opt/bitdefender-security-tools/bin/bduitool get ps
[debug] unknown field company for section ANTIVIRUS
[debug] unknown field uptodate for section ANTIVIRUS
[debug] unknown field base_version for section ANTIVIRUS
[debug] unknown field manufacturer for section ANTIVIRUS
[debug] unknown field version for section ANTIVIRUS
[debug] unknown field base_creation for section ANTIVIRUS
[debug] unknown field name for section ANTIVIRUS
Use of uninitialized value in concatenation (.) or string at /usr/share/glpi-agent/lib/GLPI/Agent/Task/Inventory/Linux/AntiVirus/Bitdefender.pm line 29.
[debug2] Added 

```

Looking forward to your indications

g-bougard commented on lib/GLPI/Agent/Task/Inventory/Linux/AntiVirus/Bitdefender.pm:
```diff
@@ -0,0 +1,71 @@
+            entry   => $antivirus
+        );
+
+        $logger->debug2("Added $antivirus->{name}" . ($antivirus->{version} ? " v$antivirus->{version}" : ""))
```
Hash values are also wrongly referenced here.
```suggestion
        $logger->debug2("Added $antivirus->{NAME}" . ($antivirus->{VERSION} ? " v$antivirus->{VERSION}" : ""))
```

g-bougard commented on lib/GLPI/Agent/Task/Inventory/Linux/AntiVirus/Bitdefender.pm:
```diff
@@ -0,0 +1,71 @@
+    my @output = getAllLines(%params)
+        or return;
+
+    my $av = {
+        "name" => 'Bitdefender Endpoint Security Tools (BEST) for Linux',
+        "manufacturer" => 'Bitdefender',
+        "company" => 'Bitdefender',
+        "uptodate" => 1,
+    };
```
Hash values are also wrongly initialized here. And we don't need "MANUFACTURER".

```suggestion
    my $av = {
        NAME     => 'Bitdefender Endpoint Security Tools (BEST) for Linux',
        COMPANY  => 'Bitdefender',
        UPTODATE => 1,
    };
```

g-bougard commented on lib/GLPI/Agent/Task/Inventory/Linux/AntiVirus/Bitdefender.pm:
```diff
@@ -0,0 +1,71 @@
+        "uptodate" => 1,
+    };
+
+    foreach my $line (@output) {
+        my ($key, $value) = $line =~ /^([^:]+):\s+(.+)$/
+            or next;
+        if ($key eq "Product version") {
+            $av->{version} = $value;
+        } elsif ($key eq "Engines version") {
+            $av->{base_version} = $value;
+        } elsif ($key eq "Antimalware status") {
+            $av->{enabled} = $value eq "On" ? 1 : 0;
+        } elsif ($key =~ /New (product update|security content) available/) {
+            # Set "uptodate" to 0 if one of "new product update available" or "new security content available" is not "no"
+            $av->{uptodate} = 0 if $value ne "no";
+        } elsif ($key eq "Last security content update" && $value =~ /^(\d{4}-\d+-\d+) at/) {
+            $av->{base_creation} = $1;
+        }
+    }
```
Hash values are also wrongly referenced here.

```suggestion
    foreach my $line (@output) {
        my ($key, $value) = $line =~ /^([^:]+):\s+(.+)$/
            or next;
        if ($key eq "Product version") {
            $av->{VERSION} = $value;
        } elsif ($key eq "Engines version") {
            $av->{BASE_VERSION} = $value;
        } elsif ($key eq "Antimalware status") {
            $av->{ENABLED} = $value eq "On" ? 1 : 0;
        } elsif ($key =~ /New (product update|security content) available/) {
            # Set "uptodate" to 0 if one of "new product update available" or "new security content available" is not "no"
            $av->{UPTODATE} = 0 if $value ne "no";
        } elsif ($key eq "Last security content update" && $value =~ /^(\d{4}-\d+-\d+) at/) {
            $av->{BASE_CREATION} = $1;
        }
    }
```

g-bougard reviewed (changes requested):
Hello Oscar,

my bad, I forgot the hash value names must be capitalized.

OscarBeiro committed:
Capitalise hash values

Co-authored-by: Guillaume Bougard <gbougard@teclib.com>

OscarBeiro committed:
Fix hash values initialisation

Co-authored-by: Guillaume Bougard <gbougard@teclib.com>

OscarBeiro committed:
More hash fixing

Co-authored-by: Guillaume Bougard <gbougard@teclib.com>

OscarBeiro committed:
Merge branch 'glpi-project:develop' into featurebitdefender

g-bougard commented on lib/GLPI/Agent/Task/Inventory/Linux/AntiVirus/Bitdefender.pm:
```diff
@@ -0,0 +1,70 @@
+}
+
+1;
+
```
Last empty line should be removed as stated by unit tests.

OscarBeiro committed:
Fix Enabled

Declared Enabled, and also set Uptodate to 0.

Enabled was not working because of the different structure.
This is a quick and dirty fix, but tweaking the regex will be better.

Signed-off-by: Oscar B. <oscar@ticgal.com>

OscarBeiro commented:
Your fixes worked like a charm :)
However, the enabled value was incorrectly parsed because of the different structure.

I have made a minor fix. I'm sure you can improve it, but it works!

![image](https://github.com/glpi-project/glpi-agent/assets/18396050/de23004c-7f88-4c9a-b35f-ca535b16db39)

![image](https://github.com/glpi-project/glpi-agent/assets/18396050/1df4e174-ca9a-4fa8-ac05-ef6c960acc98)


In my opinion, it is ready.

We can improve the Expiration date, but I would like to handle this on another PR if you don't mind. External script, inject a JSON as a parameter, documentation...  not exactly an out-of-the-box experience :) . Also, the information is needed for windows too.

Merci.

g-bougard commented on lib/GLPI/Agent/Task/Inventory/Linux/AntiVirus/Bitdefender.pm:
```diff
@@ -0,0 +1,71 @@
+            $av->{VERSION} = $value;
+        } elsif ($key eq "Engines version") {
+            $av->{BASE_VERSION} = $value;
+        } elsif ($key eq "  - Antimalware status") {
```
To follow your suggestion to update the regexp to fix enabled status:
```suggestion
        my ($key, $value) = $line =~ /^(?:\s+-\s)?([^:]+):\s+(.+)$/
            or next;
        if ($key eq "Product version") {
            $av->{VERSION} = $value;
        } elsif ($key eq "Engines version") {
            $av->{BASE_VERSION} = $value;
        } elsif ($key eq "Antimalware status") {
```

g-bougard commented on lib/GLPI/Agent/Task/Inventory/Linux/AntiVirus/Bitdefender.pm:
```diff
@@ -0,0 +1,71 @@
+        NAME     => 'Bitdefender Endpoint Security Tools (BEST) for Linux',
+        COMPANY  => 'Bitdefender',
+        ENABLED => 0,
+        UPTODATE => 0,
```
In your last commit, you changed UPTODATE default to 0. But UPTODATE flag should be kept to 1 by default as on l.61 it will be reset to 0 is one of the required field is not set. As now, it will be kept to 0 anyway ;-)
```suggestion
        UPTODATE => 1,
```

OscarBeiro replied:
OK. I thought it was better default to 0.

g-bougard commented:
Anyway, I agree to report EXPIRATION support to another dev as this is not trivial.

OscarBeiro committed:
Set Uptodate to 1

Co-authored-by: Guillaume Bougard <gbougard@teclib.com>

OscarBeiro committed:
Enhance regex

Co-authored-by: Guillaume Bougard <gbougard@teclib.com>

OscarBeiro commented:
It should be ready now. 
![image](https://github.com/glpi-project/glpi-agent/assets/18396050/afa3b45e-93e7-46d8-852a-e112852cc90a)
Big thank you!

Finally, one side note is not directly related; maybe you are already aware.
I've noticed that when a partial inventory is requested, it generates a JSON, and by default, the Linux agent generates an XML. 
![image](https://github.com/glpi-project/glpi-agent/assets/18396050/f9ff4e5d-ca5d-44fe-a981-52be630abc67)

OscarBeiro reviewed (commented):
I think it is ready

g-bougard commented:
It's not exactly ready as you need to address this failing unittest: https://github.com/glpi-project/glpi-agent/actions/runs/6087244159/job/16515399625?pr=472
It seems you just need to remove the last empty line.

OscarBeiro committed:
Remove blanks

Signed-off-by: Oscar B. <oscar@ticgal.com>

OscarBeiro commented:
> Hello Oscar,
> 
> my bad, I forgot the hash value names must be capitalized.

This request is still pending. Are you expecting me to do something? 
Merci
